### PR TITLE
[insurance-refacto] handle savecart module with insurance

### PR DESCRIPTION
**Pull request by Benjamin-Freoua-Alma**
### Reason for change

[Linear task](https://linear.app/almapay/issue/ECOM-1691/[🧩-ps]-fix-compatibility-with-a-savecart-module)

### Code changes

On save cart (if curent cart id != new cart id), on the hook saveCart, we duplicate the cart

### How to test

_As a reviewer, you are encouraged to test the PR locally._

Install a module for save cart (eg.: opartsavecart)
Save a cart with insurance and check if the insurance design product is displayed

### Checklist for authors and reviewers

- [ ] The title of the PR uses business wording, not technical jargon, for the changelog readers to understand it
- [ ] The PR implements the changes asked in the referenced task / issue
- [ ] The automated tests are compliant with the [testing strategy](https://www.notion.so/almapay/Backend-testing-strategy-06c642cec1bf47b9b8feca3a91ea8d4a)
- [ ] The tests are relevant, and cover the corner/error cases, not only the happy path
- [ ] You understand the impact of this PR on existing code/features
- [ ] The changes include adequate logging and Datadog traces
- [ ] Documentation is updated (API, developer documentation, ADR, Notion...)

### CODE REVIEW

**Review by ClaireAlma — CHANGES_REQUESTED** [review]
Hi, can we review this pr in meeting together ? Thanks

File: alma/lib/Services/CartService.php [hunk 1]
```diff
@@ -38,9 +43,90 @@ class CartService
...
+     *
+     * @return void
+     *
+     * @throws AlmaException
+     * @throws \PrestaShopException
+     */
+    public function duplicateCart($newCart)
+    {
+        $baseCart = $this->contextFactory->getContextCart();
+
```
Comment (Francois-Gomis) [inline alma/lib/Services/CartService.php:97]: Dependency injection principle - Controller inject cart to function

File: alma/lib/Services/CartService.php [hunk 2]
```diff
@@ -38,9 +43,90 @@ class CartService
...
+    /**
+     * @param \Cart $newCart
+     *
+     * @return void
+     *
+     * @throws AlmaException
+     * @throws \PrestaShopException
+     */
+    public function duplicateCart($newCart)
```
Comment (Francois-Gomis) [inline alma/lib/Services/CartService.php:94]: not the good name ?

File: alma/lib/Services/CartService.php [hunk 3]
```diff
@@ -38,9 +43,90 @@ class CartService
...
+     *
+     * @return void
+     *
+     * @throws AlmaException
+     * @throws \PrestaShopException
+     */
+    public function duplicateInsuranceProductsInDB($newCart, $currentCart)
+    {
+        if (!$this->insuranceHelper->checkInsuranceProductsExist($newCart)) {
+            try {
```
Comment (Francois-Gomis) [inline alma/lib/Services/CartService.php:122]: newCart Or Current Cart ?  - naming -  check if alreadyCreated ?  // add comment multi run

File: alma/lib/Services/InsuranceProductService.php [hunk 4]
```diff
@@ -253,9 +270,18 @@ public function addInsuranceProduct(
...
     {
         try {
+            if (!$this->canHandleAddingProductInsurance()) {
+                return false;
+            }
+
+            $idProduct = \Tools::getValue('id_product');
+            $insuranceContractId = \Tools::getValue('alma_id_insurance_contract');
+            $quantity = \Tools::getValue('qty');
+            $idCustomization = \Tools::getValue('id_customization');
```
Comment (Francois-Gomis) [inline alma/lib/Services/InsuranceProductService.php:283]: add toolFactory - dependency inversion - you can inject params to no longer depend on the execution context

File: alma/lib/Services/InsuranceProductService.php [hunk 5]
```diff
@@ -253,9 +270,18 @@ public function addInsuranceProduct(
      *
      * @return bool
      */
-    public function handleAddingProductInsurance($idProduct, $insuranceContractId, $quantity, $idCustomization, $cart = null, $destroyPost = true)
+    public function handleAddingProductInsurance($cart = null)
     {
         try {
+            if (!$this->canHandleAddingProductInsurance()) {
+                return false;
+            }
```
Comment (Francois-Gomis) [inline alma/lib/Services/InsuranceProductService.php:278]: Check in controller - Adding product execution context - check multi run ?

File: alma/lib/Services/InsuranceProductService.php [hunk 6]
```diff
@@ -217,9 +236,7 @@ public function addInsuranceProduct(
             $this->context->shop->id
         );
 
-        if ($destroyPost) {
-            $_POST['alma_id_insurance_contract'] = null;
-        }
+        $_POST['alma_id_insurance_contract'] = null;
```
Comment (Francois-Gomis) [inline alma/lib/Services/InsuranceProductService.php:239]: ??

File: alma/lib/Services/InsuranceProductService.php [hunk 7]
```diff
@@ -253,9 +270,18 @@ public function addInsuranceProduct(
      *
      * @return bool
      */
-    public function handleAddingProductInsurance($idProduct, $insuranceContractId, $quantity, $idCustomization, $cart = null, $destroyPost = true)
+    public function handleAddingProductInsurance($cart = null)
```
Comment (Francois-Gomis) [inline alma/lib/Services/InsuranceProductService.php:273]: name ?

File: alma/lib/Services/InsuranceProductService.php [hunk 8]
```diff
@@ -253,9 +270,18 @@ public function addInsuranceProduct(
      *
      * @return bool
```
Comment (Francois-Gomis) [inline alma/lib/Services/InsuranceProductService.php:271]: return is not used  change to void and use exception

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 9]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
+        $moduleMock = \Mockery::mock(\Module::class)->makePartial();
+        $moduleMock->name = ConstantsHelper::ALMA_MODULE_NAME;
+        $moduleFactoryMock->shouldReceive('getModule')->andReturn($moduleMock);
+
         $this->assertEquals(ConstantsHelper::ALMA_MODULE_NAME, $this->moduleFactory->getModuleName());
     }
 
     public function testL()
     {
         $this->assertEquals('Pay now by credit card', $this->moduleFactory->l('Pay now by credit card', ConstantsHelper::SOURCE_CUSTOM_FIELDS));
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Factories/ModuleFactoryTest.php:91]: translation not working in test , what is the point of the test ?

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 10]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
     }
+
+    public function testIsInstalledPsAfter17()
+    {
+        $this->toolsHelperMock->shouldReceive('psVersionCompare')->andReturn(false);
+
+        $moduleFactory = \Mockery::spy(ModuleFactory::class, [$this->toolsHelperMock])->makePartial();
+        $moduleFactory->shouldReceive('isInstalledAfter17');
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Factories/ModuleFactoryTest.php:99]: not necessary

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 11]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
+        $moduleFactory->shouldHaveReceived('isInstalledAfter17');
+        $moduleFactory->shouldNotHaveReceived('isInstalledBefore17');
+    }
+
+    public function testIsInstalledPsBefore17()
+    {
+        $this->toolsHelperMock->shouldReceive('psVersionCompare')->andReturn(true);
+
+        $moduleFactory = \Mockery::spy(ModuleFactory::class, [$this->toolsHelperMock])->makePartial();
+        $moduleFactory->shouldReceive('isInstalledBefore17');
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Factories/ModuleFactoryTest.php:112]: not necessary

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 12]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
+    public function testIsInstalledPsAfter17()
+    {
+        $this->toolsHelperMock->shouldReceive('psVersionCompare')->andReturn(false);
+
+        $moduleFactory = \Mockery::spy(ModuleFactory::class, [$this->toolsHelperMock])->makePartial();
+        $moduleFactory->shouldReceive('isInstalledAfter17');
+
+        $moduleFactory->isInstalled('alma');
+
+        $moduleFactory->shouldHaveReceived('isInstalledAfter17');
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Factories/ModuleFactoryTest.php:103]: you can add with('alma')

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 13]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
+    public function testIsInstalledPsBefore17()
+    {
+        $this->toolsHelperMock->shouldReceive('psVersionCompare')->andReturn(true);
+
+        $moduleFactory = \Mockery::spy(ModuleFactory::class, [$this->toolsHelperMock])->makePartial();
+        $moduleFactory->shouldReceive('isInstalledBefore17');
+
+        $moduleFactory->isInstalled('fakename');
+
+        $moduleFactory->shouldHaveReceived('isInstalledBefore17');
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Factories/ModuleFactoryTest.php:116]: you can add with('fakename')

File: alma/tests/Unit/Helper/InsuranceProductHelperTest.php [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+        $this->almaInsuranceProductRepositorySpy = \Mockery::spy(AlmaInsuranceProductRepository::class)->makePartial();
+
+        $this->almaInsuranceProductRepositorySpy->shouldReceive('getByCartIdAndShop')
+            ->with($this->cart->id, $this->context->shop->id)
+            ->andReturn($almaInsuranceProduct);
+
+        $this->almaInsuranceProductRepositorySpy->shouldReceive('add')->andReturn(null);
+
+        $this->insuranceProductHelper = \Mockery::mock(InsuranceProductHelper::class, [
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Helper/InsuranceProductHelperTest.php:101]: you should not make a mock if it is not necessary for the class on which you are testing

File: alma/lib/Services/InsuranceProductService.php (hunk 6 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Services/InsuranceProductService.php:243]: I think it was for PS 1.6. I remove it and we will tested it :)

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 10 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php:99]: without makePartial, the test doesn't work :/

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 11 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php:112]: without makePartial, the test doesn't work :/

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 12 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: I can only use Mokery in this test

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 13 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php:116]: I can only use Mokery in this test

**Review by Francois-Gomis — CHANGES_REQUESTED** [review]
Change optSaveCart/CartService name for php 5.6 compatibility

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 12 above)
Comment (Francois-Gomis) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: You can test used parameter with ->with()

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 11 above)
Comment (Francois-Gomis) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: Ping me when you change - a try and that work

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 9 above)
Comment (Francois-Gomis) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: ??

File: alma/tests/Unit/Factories/ModuleFactoryTest.php [hunk 15]
```diff
@@ -35,23 +37,83 @@ class ModuleFactoryTest extends TestCase
...
     {
         $this->assertInstanceOf(\Module::class, $this->moduleFactory->getModule());
     }
 
-    public function testGetContextName()
+    public function testGetModuleNameNoModule()
+    {
+        $moduleFactoryMock = \Mockery::mock(ModuleFactory::class, [$this->toolsHelperMock])->makePartial();
+        $moduleFactoryMock->shouldReceive('getModule')->andReturn(false);
+        $this->assertEquals('', $moduleFactoryMock->getModule());
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: You tell me when I use internal method I must use `makePartial()`
We can speak on it if I don't understand this issue :)

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 9 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: I forgot to update it 🙏

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 11 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: We'll see together in synch?

File: alma/tests/Unit/Factories/ModuleFactoryTest.php (hunk 12 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Factories/ModuleFactoryTest.php]: like this ?

**Review by Francois-Gomis — APPROVED** [review]

**Review by joyet-simon — APPROVED** [review]

**Review by carine-bonnafous — APPROVED** [review]
(DevX review was triggered only because of a line removed in .github/workflows/e2e-test-post-results.yml)

**Pull request merged**